### PR TITLE
Verify regions

### LINKED ISSUE #963: Verify register kinds
Enhance bir/verify.bal to verify that the kind of a BIR register is consistent with how it is used within the FunctionCode.
Follow-up from #920

### LINKED ISSUE #995: Verify register kinds and regions in BIR
**Description:**

Verify register kinds and regions in the BIR verfier

- [x] #963

Check if
- Temp registers are assigned once and only once as a result of an instruction.
- List/map operand in ListSetInsn / MappingSetInsn is not a final register.
- Temp registered are assigned in a previous instruction in the block before being used as a operand in another instruction.
- Narrow register is a subtype of underlying unnarrowed register.

- [ ] Verify regions

Check if
- All blocks in the function code can be reached from block 0 via forward edges
- Backward branches must point to a block that is a loop head
- Region records are created for each single-entry single-exit region in the function code
- Graphs formed by forward edges must be acyclic

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

### PR DESCRIPTION
**Pull request by ushirask**
Partially fixes #995

### CODE REVIEW

File: compiler/modules/bir/verify.bal [hunk 1]
```diff
@@ -79,6 +80,54 @@ type Error err:Semantic|err:Internal;
...
+    else if insn is CallInsn {
+        foreach Operand arg in insn.args {
+            check verifyRegisterKind(vc, arg);
+        }
+    }
+    else if insn is ListSetInsn|MappingSetInsn {
+        check verifyNonFinalRegisterKind(vc, insn.operands[0]);
+    }
+    else if insn is MultipleOpeandInsn {
+        foreach var op in <Operand[]>insn.operands {
```
Comment (ushirask) [inline compiler/modules/bir/verify.bal]: @jclark without the type cast it throws the error 
`ERROR [verify.bal:(125:27,125:40)] incompatible types: '((wso2/nballerina.bir:0.1.0:IntOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:IntOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:IntOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:FloatOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:DecimalOperand[2] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[2] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[] & readonly)|([wso2/nballerina.bir:0.1.0:Register,wso2/nballerina.bir:0.1.0:IntOperand] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[] & readonly)|([wso2/nballerina.bir:0.1.0:Register,wso2/nballerina.bir:0.1.0:StringOperand] & readonly)|(wso2/nballerina.bir:0.1.0:StringOperand[2] & readonly))' is not an iterable collection`

This seems independent of the insn type. Even if we use just 2 instructions with same `operands` it still fails.
For `type MultipleOpeandInsn ListConstructInsn|MappingConstructInsn;` 
`incompatible types: '((wso2/nballerina.bir:0.1.0:Operand[] & readonly)|(wso2/nballerina.bir:0.1.0:Operand[] & readonly))' is not an iterable collection`

**Comment by manuranga** [comment]
Hi @jclark 
Should we verify that a give tmp variables is used ones and only ones **in a given branch**? I think we do. If so, Ushira will need to take control flow into account.

File: compiler/modules/bir/verify.bal (hunk 1 above)
Comment (ushirask) [reply compiler/modules/bir/verify.bal]: Added an issue for JBUG in https://github.com/ballerina-platform/ballerina-lang/issues/35557

**Comment by jclark** [comment]
> Hi @jclark Should we verify that a give tmp variables is used ones and only ones **in a given branch**? I think we do. If so, Ushira will need to take control flow into account.

For tmp, I was envisaging that we would check that it occurs as the result of exactly one instruction.

File: compiler/modules/bir/verify.bal (hunk 1 above)
Comment (jclark) [reply compiler/modules/bir/verify.bal]: Can we work around by assigning to a variable before iterating?

Comment (ushirask) [reply compiler/modules/bir/verify.bal]: Yes, changed it to assign and iterate.

File: compiler/modules/bir/verify.bal [hunk 2]
```diff
@@ -90,22 +90,27 @@ type MultipleOpeandInsn IntBinaryInsn|IntNoPanicArithmeticBinaryInsn|FloatArithm
     |MappingGetInsn|StringConcatInsn;
 
 type NonResultInsn MappingSetInsn|ListSetInsn|BranchInsn|CondBranchInsn|CondNarrowInsn|PanicInsn
-    |RetInsn|AbnormalRetInsn;
+    |RetInsn|AbnormalRetInsn|AssignInsn;
 
 function verifyRegistersKinds(VerifyContext vc, Insn insn) returns Error? {
     if insn !is NonResultInsn {
-        vc.definedTmpRegisters.push(insn.result.number);
+        if vc.definedTmpRegisters.indexOf(insn.result.number, 0) != () {
```
Comment (manuranga) [inline compiler/modules/bir/verify.bal]: No need for `0` its the defualt

File: compiler/modules/bir/verify.bal [hunk 3]
```diff
@@ -79,6 +81,84 @@ type Error err:Semantic|err:Internal;
...
+        }
+    }
+    else {
+        check verifyRegisterKind(vc, insn.operand);
+    }
+}
+
+function verifyNonFinalRegisterKind(VerifyContext vc, Operand r) returns Error? {
+    if r is FinalRegister {
+            return vc.invalidErr("invalid register kind final for insn: " + r.kind, <Position>r.pos);
```
Comment (manuranga) [inline compiler/modules/bir/verify.bal]: too many spaces in indention.

File: compiler/modules/bir/verify.bal (hunk 2 above)
Comment (ushirask) [reply compiler/modules/bir/verify.bal]: Fixed.

File: compiler/modules/bir/verify.bal [hunk 4]
```diff
@@ -79,6 +82,71 @@ type Error err:Semantic|err:Internal;
 function verifyBasicBlock(VerifyContext vc, BasicBlock bb) returns Error? {
     foreach Insn insn in bb.insns {
         check verifyInsn(vc, insn);
+        check verifyRegistersKinds(vc, insn);
+    }
+}
+
+type MultipleOpeandInsn IntBinaryInsn|IntNoPanicArithmeticBinaryInsn|FloatArithmeticBinaryInsn
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: Spelling

File: compiler/modules/bir/verify.bal [hunk 5]
```diff
@@ -79,6 +82,71 @@ type Error err:Semantic|err:Internal;
...
+    |DecimalArithmeticBinaryInsn|CompareInsn|ListConstructInsn|ListGetInsn|MappingConstructInsn
+    |MappingGetInsn|StringConcatInsn|EqualityInsn;
+
+type NonResultInsn MappingSetInsn|ListSetInsn|BranchInsn|CondBranchInsn|CondNarrowInsn|PanicInsn
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: Cleaner to do it the other way round: `ResultInsn`.

File: compiler/modules/bir/verify.bal [hunk 6]
```diff
@@ -10,12 +10,15 @@ class VerifyContext {
     private final Module mod;
     private final t:Context tc;
     private final FunctionDefn defn;
-
-    function init(Module mod, FunctionDefn defn) {
+    final FunctionCode code;
+    int[] definedTmpRegisters = [];
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: `boolean[] tmpRegisterUsed`

File: compiler/modules/bir/verify.bal [hunk 7]
```diff
@@ -79,6 +82,71 @@ type Error err:Semantic|err:Internal;
...
+        return vc.invalidErr("invalid register kind final for insn: " + r.kind, <Position>r.pos);
+    }
+    check verifyRegisterKind(vc, r);
+}
+
+function verifyRegisterKind(VerifyContext vc, Operand r) returns Error? {
+    if r !is Register {
+        return;
+    }
+    if r is TmpRegister && vc.definedTmpRegisters.indexOf(r.number) == () {
```
Comment (jclark) [inline compiler/modules/bir/verify.bal]: Don't want to make assumptions here about ordering of blocks.

**Review by jclark — COMMENTED** [review]
First of all, in the issue write a precise description of everything that  you are checking.

**Comment by jclark** [comment]
I can't review this until the description of the problem in #995 is fixed.

**Review by jclark — CHANGES_REQUESTED** [review]
Please fix the problems with the issue description in #995 first.

**Pull request closed without merging**